tools/docker_memory_check.py

```diff
@@ -0,0 +1,72 @@
+from __future__ import annotations
+
+import argparse
+import json
+import os.path
+import sys
+from typing import Sequence
+
+
+def should_use_color(setting: str) -> bool:
+    # normally I would use `sys.stdout.isatty()` however direnv always pipes this
+    return setting == "always" or (setting == "auto" and not os.environ.get("CI"))
+
+
+def color(s: str, color: str, *, use_color: bool) -> str:
+    if use_color:
+        return f"{color}{s}\033[m"
+    else:
+        return s
+
+
+def main(argv: Sequence[str] | None = None) -> int:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--settings-file",
+        default=os.path.expanduser("~/Library/Group Containers/group.com.docker/settings.json"),
+        help=argparse.SUPPRESS,
+    )
+    parser.add_argument(
+        "--memory-minimum",
+        default=8092,
+        type=int,
+        help="the minimum amount of allocated memory to warn for.  default: %(default)s (MiB)",
+    )
+    parser.add_argument(
+        "--color",
+        choices=("always", "never", "auto"),
+        default="auto",
+        help="whether to use color.  default: %(default)s (auto is determined by CI environment variable)",
+    )
+    args = parser.parse_args(argv)
+
+    use_color = should_use_color(args.color)
+
+    try:
+        with open(args.settings_file) as f:
+            contents = json.load(f)
+    except (json.JSONDecodeError, OSError):
+        return 0  # file didn't exist or was not json
```

Comment:
I know it is not quite the exit code since main is wrap inside SystemExit, but I wonder why returning 0? Doesn't this make the main function always return 0?

It's weird that it doesn't fail and bad/good inputs end with the same exit code:
```
python tools/docker_memory_check.py --settings-file foo.json && echo $? || echo $?
0

python tools/docker_memory_check.py --settings-file "~/Library/Group Containers/group.com.docker/settings.json" && echo $? || echo $?
0
```

Reply:
this is the script exitcode!  the `SystemExit` pattern allows it to exit and be tested.

- testable command lines: https://youtu.be/sv46294LoP8
- why `SystemExit`: https://youtu.be/ZbeSPc5wL0g

as for why it's always zero here: a nonzero exit code would break `.envrc` -- and for a warning I decided that it should pass "silently"

Reply:
We can probably use `||` in `.envrc` or something similar.

Reply:
if we're always ignoring the exit code there's probably not much reason to set a special one

btw instead of `&& echo $? || echo $?` you can do `; echo $?` (at least without `-e` set)